x/ibc/04-channel/keeper/handshake.go

```diff
@@ -0,0 +1,368 @@
+package keeper
+
+import (
+	"errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	ics03types "github.com/cosmos/cosmos-sdk/x/ibc/03-connection/types"
+	"github.com/cosmos/cosmos-sdk/x/ibc/04-channel/types"
+	ics23 "github.com/cosmos/cosmos-sdk/x/ibc/23-commitment"
+)
+
+// ChanOpenInit is called by a module to initiate a channel opening handshake with
+// a module on another chain.
+func (k Keeper) ChanOpenInit(
+	ctx sdk.Context,
+	order types.ChannelOrder,
+	connectionHops []string,
+	portID,
+	channelID string,
+	counterparty types.Counterparty,
+	version string,
+) (string, error) {
+	// TODO: abortTransactionUnless(validateChannelIdentifier(portIdentifier, channelIdentifier))
+	if len(connectionHops) != 1 {
+		return "", types.ErrInvalidConnectionHops(k.codespace)
+	}
+
+	_, found := k.GetChannel(ctx, portID, channelID)
+	if found {
+		return "", types.ErrChannelExists(k.codespace)
+	}
+
+	connection, found := k.connectionKeeper.GetConnection(ctx, connectionHops[0])
+	if !found {
+		return "", ics03types.ErrConnectionNotFound(k.codespace)
+	}
+
+	// TODO: inconsistency on ICS03 (`none`) and ICS04 (`CLOSED`)
+	if connection.State == ics03types.NONE {
+		return "", errors.New("connection is closed")
+	}
+
+	// TODO: Blocked - ICS05 Not implemented yet
+	// port, found := k.portKeeper.GetPort(ctx, portID)
+	// if !found {
+	// 	return errors.New("port not found") // TODO: ics05 sdk.Error
+	// }
+
+	// if !k.portKeeper.AuthenticatePort(port.ID()) {
+	// 	return errors.New("port is not valid") // TODO: ics05 sdk.Error
+	// }
+
+	channel := types.NewChannel(types.INIT, order, counterparty, connectionHops, version)
+	k.SetChannel(ctx, portID, channelID, channel)
+
+	key := "" // TODO: generate key
```

Comment:
What is required for key generation here?

Reply:
This should be identical to the logic for store keys.